### DISCUSSION TITLE
Frontmatter redundancy

### DISCUSSION THREAD

**Discussion by aalexei** [protesilaos/denote#355] [General]
I'm transitioning from org-roam to denote but before I convert a whole bunch of notes, will I end up with problems if my frontmatter only includes the #+title: and some other custom fields? 

The default at the moment also adds a #+date, #+identifier, and #+filetags but that seems redundant since all that info in is the filename. It also introduces friction for me, since if they become different I then have to decide which was correct, so I'd rather remove them and have a single source of that info. I'm worried that if I omit them some functionality will break somewhere.

**Comment by ZharMeny** [comment]
Frontmatter is the metadata, and probably should be the 'source of truth', while the filename is for sorting/filename searching.  You can also synchronise the filename with the frontmatter using `denote-rename-file-using-front-matter`.  I would imagine most functionality should not break unless you remove the note's identifier.  Note that the ID of the note uses the note creation time (in ISO 8601-like format).

P.S. As you are transitioning from org-roam I should probably also mention the `denote-date` command, which will prompt for the DATE and TIME you want to create a note with, along the usual TITLE and KEYWORDS.

**Comment by aalexei** [comment]
The problem with the front matter as the source of truth is that not all notes have it by design. One of the great advantages of denotes filename scheme is that binary notes are first class citizens, I can add non-text notes and tag, search and link to them.  The file name is the more 'primitive' object also, and accessible without even having to open and parse the files, so at a pinch I can throw together a simple bash script to process my notes without worrying about content format. As far as I can see from the code, `denote-id-regexp` is only used to parse the id out of the filename, not the frontmatter. In fact `denote-rename-file-using-front-matter` never touches the id and the function has the comment (I think this answers my OP!):
```
Never modify the identifier of the FILE, if any, even if it is
edited in the front matter.  Denote considers the file name to be
the source of truth in this case to avoid potential breakage with
typos and the like.
```

In contrast, the title in the filename is a sluggified version of the 'true' title in the frontmatter when it's a text note. Buffer listing go off the true title and `denote-link` will insert the true title. The #+date seems to carry the same info as the id, not sure if there are some org functions that use it? There *are* independent org function to set file-tags and you could use  them together with `denote-rename-file-using-front-matter` to propagate them to the file name, but it doesn't seem much of a loss to just use denotes keyword functions.

**Comment by aalexei** [comment]
Incidentally, just tried  `denote-rename-file-using-front-matter`/`denote-keywords-add` etc with 
```
#+title:      Test node. With functions
:drawer:
#+date:       [2024-05-05 Sun 17:13]
#+property: DIR 20240505T171333
#+author: Blah Blah Blah
#+subtitle: Blah Blah
#+startup: latexpreview
#+setupfile: ~/setupfile.org
#+filetags:   :denote:test:
#+identifier: 20240505T171333
:end:
```
and it happily picked out the filetags to update despite all the header noise and the drawer. This means you can put everything in an org drawer and collapse it and get a visually clean note, and not lose any functionality. Cool.

**Comment by Nebucatnetzer** [comment]
I feel it is a bit backwards, that the front matter is the source of truth.
As mentioned denote is file agnostic and the `filetags` that get added don't really need to have something to do with the actual org tags used in the file itself which then causes problems if you use tag inheritance for example.

**Comment by protesilaos** [comment]
Hello folks! I missed the previous notifications related to this thread. I do not know what happened. Anyway, I got the latest from @Nebucatnetzer.

> I feel it is a bit backwards, that the front matter is the source of truth.

In general, the source of truth is the file name itself. It is possible to not have any front matter, for example.

Where the front matter is the input is in the commands `denote-rename-file-using-front-matter` and `denote-dired-rename-marked-files-using-front-matter`.

If there is a case where this is not true, then we need to look into it: it is not intentional.

**Reply by Nebucatnetzer** [reply]
That is good to know, thank you.
A function like `denote-recreate-front-matter` which takes the information from the filename would be cool in that case.
Because only the other way exists the filename felt less official to me.

**Reply by protesilaos** [reply]
Something like this should work. Though it does not convert the title into something prettier:

```elisp
(defun denote-recreate-front-matter (file)
  "Recreate the front matter of FILE."
  (interactive (list (denote--rename-dired-file-or-current-file-or-prompt)))
  (unless (denote-file-is-writable-and-supported-p file)
    (user-error "The `%s' file cannot have front matter" file))
  (denote-rename-file file 'keep-current 'keep-current 'keep-current 'keep-current 'keep-current))
```

What is the use-case for this though?

In my case, I have many text files without any front matter because I do not need it. The front matter is there for the convenience of the user when, say, they want some other Org package/functionality to read the data.

**Reply by Nebucatnetzer** [reply]
I usually create all my notes with denote and set the front matter on creation because some of them are nice to have in each file (title, author, setupfile, latex_header, date).
Sometimes I then change the filename in a GUI file manager or the terminal so it would be handy to be able to e.g. update the front matter.
Or I have some old notes and I would like to migrate them to denote.
`denote-rename-file` is close but it is a bit involved if you don't actually want to rename the file.

**Comment by protesilaos** [comment]
An aside here, I feel this separate "discussions" page is confusing to me. What is the difference with the issue tracker? I will need to look into it and probably move everything to the issue tracker.

**Reply by Nebucatnetzer** [reply]
Often developers use the issue tracker only for actual bugs or features they want to develop.
So a discussion can discuss a topic and potentially identify a bug which then becomes and issue.
In addition you can use discussions as a dynamic FAQ and mark answers as the solution to a question similar to Stackoverflow.
But I agree that it can be a bit confusing to have both places.

**Reply by protesilaos** [reply]
I see, thank you! Let's keep it then.
